### PR TITLE
TEAM/MHBA: Adds mkey cache

### PR DESCRIPTION
**Pull request by vspetrov**
ASR may skip mkey update step (for SEND or RECV) if the following
    condition is met:
	1. registration has not changed
	2. Buffer pointer has not changed
	3. Msg size has not changed

Alternative to #104

### CODE REVIEW

File: src/team_lib/mhba/xccl_mhba_collective.c [hunk 1]
```diff
@@ -63,34 +69,51 @@ static xccl_status_t xccl_mhba_reg_fanin_start(xccl_coll_task_t *task)
     xccl_mhba_task_t     *self    = ucs_derived_of(task, xccl_mhba_task_t);
     xccl_mhba_coll_req_t *request = self->req;
     xccl_mhba_team_t     *team    = request->team;
-    int                   sr_mem_access_flags = 0;
-    int dr_mem_access_flags = IBV_ACCESS_LOCAL_WRITE | IBV_ACCESS_REMOTE_WRITE;
-
+    int                   index   = SEQ_INDEX(request->seq_num);
+    ucs_rcache_region_t*  send_ptr;
+    ucs_rcache_region_t*  recv_ptr;
+    int                   reg_change_flag;
```
Comment (lappazos) [inline src/team_lib/mhba/xccl_mhba_collective.c]: set to 0

File: src/team_lib/mhba/xccl_mhba_collective.c [hunk 2]
```diff
@@ -63,34 +69,51 @@ static xccl_status_t xccl_mhba_reg_fanin_start(xccl_coll_task_t *task)
...
     }
     request->send_rcache_region_p = xccl_rcache_ucs_get_reg_data(send_ptr);
 
+    /* NOTE: we does not support alternating block_size for the same msg size - TODO
+       Will need to add the possibility of block_size change into consideration
+       when initializing the mkey_cache_flag */
+
+    xccl_mhba_get_my_ctrl(team, index)->mkey_cache_flag =
+        (request->args.buffer_info.len == team->previous_msg_size[index]) ? 0 :
+        (XCCL_MHBA_NEED_RECV_MKEY_UPDATE | XCCL_MHBA_NEED_RECV_MKEY_UPDATE);
```
Comment (lappazos) [inline src/team_lib/mhba/xccl_mhba_collective.c:92]: needed only for asr - i know its not significant in performance, but still

File: src/team_lib/mhba/xccl_mhba_collective.c [hunk 3]
```diff
@@ -63,34 +69,51 @@ static xccl_status_t xccl_mhba_reg_fanin_start(xccl_coll_task_t *task)
...
+
+    if (reg_change_flag || (request->send_rcache_region_p->mr->addr !=
+                            team->previous_send_address[index])) {
+        xccl_mhba_get_my_ctrl(team, index)->mkey_cache_flag |= XCCL_MHBA_NEED_SEND_MKEY_UPDATE;
+    }
+
     if(UCS_OK != ucs_rcache_get(team->rcache, (void *)request->args.buffer_info.dst_buffer,
                                 request->args.buffer_info.len * team->size,
-                                PROT_WRITE,&dr_mem_access_flags,&recv_ptr)) {
+                                PROT_WRITE, &reg_change_flag, &recv_ptr)) {
```
Comment (lappazos) [inline src/team_lib/mhba/xccl_mhba_collective.c:101]: need additional set to 0

File: src/team_lib/mhba/xccl_mhba_collective.c [hunk 4]
```diff
@@ -547,7 +571,6 @@ xccl_status_t xccl_mhba_alltoall_init(xccl_coll_op_args_t  *coll_args,
     request->seq_num = team->sequence_number;
     xccl_mhba_debug("Seq num is %d", request->seq_num);
     team->sequence_number++;
-
     for (i = 0; i < n_tasks; i++) {
         request->tasks[i].req = request;
         xccl_coll_task_init(&request->tasks[i].super);
```
Comment (lappazos) [inline src/team_lib/mhba/xccl_mhba_collective.c:576]: need to reset mkey_cache_flag + add check to reset seq_num if reached MAX

File: src/team_lib/mhba/xccl_mhba_collective.c (hunk 1 above)
Comment (vspetrov) [reply src/team_lib/mhba/xccl_mhba_collective.c]: yes

File: src/team_lib/mhba/xccl_mhba_collective.c (hunk 2 above)
Comment (vspetrov) [reply src/team_lib/mhba/xccl_mhba_collective.c:92]: i don't want to add extra branch and it is more clear when all the checks are in the same place. Basically a NON-asr rank reports: my former mkey is not valid.

File: src/team_lib/mhba/xccl_mhba_collective.c (hunk 3 above)
Comment (vspetrov) [reply src/team_lib/mhba/xccl_mhba_collective.c:101]: right

File: src/team_lib/mhba/xccl_mhba_collective.c (hunk 2 above)
Comment (lappazos) [reply src/team_lib/mhba/xccl_mhba_collective.c:92]: so maybe just add small todo there, maybe in future well move it somewhere, just as an FYI

File: src/team_lib/mhba/xccl_mhba_collective.c (hunk 4 above)
Comment (vspetrov) [reply src/team_lib/mhba/xccl_mhba_collective.c:576]: don't need to: mkey_cache_flag is initialized each time (former state is not re-used) so, not need to reset to 0.
and in wraparound is find, it'll start from 0.

Comment (lappazos) [reply src/team_lib/mhba/xccl_mhba_collective.c:576]: regarding seq_num check, look in 104

**Pull request merged**